futures-util/src/future/shared.rs

```diff
@@ -156,153 +203,76 @@ impl<F> Future for Shared<F>
                 }
             }
 
-            let _reset = Reset(&self.inner.notifier.state);
+            let _reset = Reset(&this.inner.notifier.state);
 
             // Poll the future
             let res = unsafe {
-                let (ref mut f, ref mut data) = *(*self.inner.future.get()).as_mut().unwrap();
-                let waker = Waker::from(self.inner.notifier.clone());
-                let mut cx = task::Context::new(data, &waker, cx.executor());
-                f.poll(&mut cx)
+                if let FutureOrOutput::Future(future) = &mut *this.inner.future_or_output.get() {
+                    PinMut::new_unchecked(future).poll(&mut cx)
+                } else {
+                    unreachable!()
+                }
             };
             match res {
-                Ok(Async::Pending) => {
+                Poll::Pending => {
                     // Not ready, try to release the handle
-                    match self.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                    match this.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
                         POLLING => {
                             // Success
-                            return Ok(Async::Pending);
+                            return Poll::Pending;
                         }
                         REPOLL => {
                             // Gotta poll again!
-                            let prev = self.inner.notifier.state.swap(POLLING, SeqCst);
+                            let prev = this.inner.notifier.state.swap(POLLING, SeqCst);
                             assert_eq!(prev, REPOLL);
                         }
                         _ => unreachable!(),
                     }
-
                 }
-                Ok(Async::Ready(i)) => {
+                Poll::Ready(i) => {
                     unsafe {
-                        (*self.inner.result.get()) = Some(Ok(SharedItem { item: Arc::new(i) }));
+                        *this.inner.future_or_output.get() = FutureOrOutput::Output(Arc::new(i));
                     }
-
-                    break;
-                }
-                Err(e) => {
-                    unsafe {
-                        (*self.inner.result.get()) = Some(Err(SharedError { error: Arc::new(e) }));
-                    }
-
                     break;
                 }
             }
         }
 
-        self.complete();
-        unsafe { self.clone_result().map(Async::Ready) }
+        unsafe {
+            this.complete();
```

Comment:
Why is the `this.complete()` call here and not after line 234? (After setting `this.inner.future_or_output`)

Reply:
Ah, took another look and realized that the result is the same. I'd still change it, though. Using `break` seems unintuitive.

Reply:
Done-- I agree, that reads much better, thanks!